sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ChannelReaderExtensions.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Channels;
+using System.Threading.Tasks;
+
+namespace Azure.Messaging.EventHubs.Core
+{
+    /// <summary>
+    ///   The set of extensions for the <see cref="ChannelReader{T}" />
+    ///   class.
+    /// </summary>
+    ///
+    internal static class ChannelReaderExtensions
+    {
+        /// <summary>The maximum delay allowed when interacting with cancellation tokens.</summary>
+        private static readonly TimeSpan maximumCancellationDelay = TimeSpan.FromDays(10);
+
+        /// <summary>
+        ///   Enumerates the events as they become available in the associated channel.
+        /// </summary>
+        ///
+        /// <typeparam name="T">The type of data contained in the channel.</typeparam>
+        ///
+        /// <param name="reader">The <see cref="ChannelReader{T}" /> instance that this method was invoked on, and from which events are to be sourced.</param>
+        /// <param name="maximumWaitTime">The maximum amount of time to wait to for an event to be available before emitting an empty item; if <c>null</c>, empty items will not be emitted.</param>
+        /// <param name="cancellationToken">The <see cref="System.Threading.CancellationToken"/> instance to signal the request to cancel enumeration.</param>
+        ///
+        /// <returns>An asynchronous enumerator that can be used to iterate over events as they are available.</returns>
+        ///
+        public static async IAsyncEnumerable<T> EnumerateChannel<T>(this ChannelReader<T> reader,
+                                                                    TimeSpan? maximumWaitTime,
+                                                                    [EnumeratorCancellation]CancellationToken cancellationToken)
+        {
+            Guard.ArgumentNotNull(nameof(reader), reader);
+
+            if (maximumWaitTime.HasValue)
+            {
+                Guard.ArgumentNotNegative(nameof(maximumWaitTime), maximumWaitTime.Value);
+            }
+
+            T result;
+
+            var waitToken = cancellationToken;
+            var waitSource = default(CancellationTokenSource);
+
+            try
+            {
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    if (reader.TryRead(out result))
+                    {
+                        waitSource?.CancelAfter(maximumCancellationDelay);
```

Comment:
Don't we have to call reset with a max value here to disable the cancellation and prevent the source from expiring while the customer is processing the event?

Reply:
I'm not sure that I see what you're picking up on.   Live 57 resets to the maximum (10 days) right before 58 yields to the caller.    Am I  overlooking something?

Reply:
Is there a reason why it should be `maximumCancellationDelay` in particular? `Timeout.Infinite` is a special value that would detach timer from timer queue freeing a bit of resources.

Reply:
I tried a bunch of different values with `CancelAfter` and hit some validation inside the library itself.   However, I don't remember trying `-1` and seeing if it was accepted and what the net effect is.   I'll run a quick prototype and see if we can go that way.   I'd prefer the built-in to a custom value.

Reply:
In some prototyping, looks like `Timeout.Infinite` is accepted and honored.   I'll put out a patch changing over to that.  Good call.